### PR TITLE
Fix MQTT5 Connect Reason Codes

### PR DESCRIPTION
**Pull request by sophokles73**
MQTT5 defines new reason codes to be included in CONNACK packets when connection establishment fails. The abstract adapter base class has been changed accordingly.

Also added integration tests based on HiveMQ client for testing connection establishment.

Addresses #2955

### CODE REVIEW

**Comment by sophokles73** [comment]
@mattkaem would you mind taking a look?

**Comment by mattkaem** [comment]
> @mattkaem would you mind taking a look?

sure, I will have a look

File: tests/src/test/java/org/eclipse/hono/tests/mqtt5/MqttTestBase.java [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+    /**
+     * Creates default MQTT client TLS options.
+     *
+     * @throws Exception if the trust manager factory cannot be created.
+     */
+    @BeforeAll
+    public static void init() throws Exception {
+
+        trustManagerFactory = new PemTrustOptions()
+                .addCertPath(IntegrationTestSupport.TRUST_STORE_PATH)
```
Comment (mattkaem) [inline tests/src/test/java/org/eclipse/hono/tests/mqtt5/MqttTestBase.java:85]: When I try to execute the integration tests I get the following error regarding this line:
```
java.lang.NullPointerException: No null certificate accepted

	at java.base/java.util.Objects.requireNonNull(Objects.java:235)
	at io.vertx.core.net.PemTrustOptions.addCertPath(PemTrustOptions.java:130)
	at org.eclipse.hono.tests.mqtt5.MqttTestBase.init(MqttTestBase.java:85)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at io.vertx.junit5.VertxExtension.interceptBeforeAllMethod(VertxExtension.java:130)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
```

I am guessing you have tested the integration tests on your end, so this is most likely a problem on my end. Have you encountered this issue before?

File: adapters/mqtt-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java [hunk 2]
```diff
@@ -1126,9 +1131,39 @@ private static MqttConnectReturnCode getConnectReturnCode(final Throwable e) {
...
+        if (e instanceof MqttConnectionException connectionException) {
+            return connectionException.code();
+        }
+        if (e instanceof AdapterConnectionsExceededException) {
+            return MqttConnectReturnCode.CONNECTION_REFUSED_USE_ANOTHER_SERVER;
+        }
+        if (e instanceof TenantConnectionsExceededException
+                || e instanceof DataVolumeExceededException
+                || e instanceof ConnectionDurationExceededException) {
+            return MqttConnectReturnCode.CONNECTION_REFUSED_QUOTA_EXCEEDED;
```
[the remark this answers is not in the document]
Comment (sophokles73) [reply adapters/mqtt-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java:1150]: Yes, MQTT 3 only defines a few return codes for hte CONNACK packet to use for indicating the reason for the failed connection attempt. In MQTT 5 there are over a dozen of _reason codes_ to choose from. This one makes the most sense IMHO ...

File: tests/src/test/java/org/eclipse/hono/tests/mqtt5/MqttTestBase.java (hunk 1 above)
Comment (sophokles73) [reply tests/src/test/java/org/eclipse/hono/tests/mqtt5/MqttTestBase.java:85]: Yes, the integration tests work on my machine ;-)
Did you build all of Hono before running the integration tests? The (test) certificates need to be packaged and installed to your local Maven repository before running the tests ...

**Comment by sophokles73** [comment]
@mattkaem any additional questions?

File: tests/src/test/java/org/eclipse/hono/tests/mqtt5/MqttTestBase.java (hunk 1 above)
Comment (mattkaem) [reply tests/src/test/java/org/eclipse/hono/tests/mqtt5/MqttTestBase.java:85]: Ok, I got the tests to run successfully on my machine now :)

**Review by mattkaem — APPROVED** [review]
LGTM :) Thanks for the contribution

**Pull request merged**